cpp/src/interop/arrow_data_structures.cpp

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/interop.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <nanoarrow/nanoarrow.h>
+#include <nanoarrow/nanoarrow.hpp>
+#include <nanoarrow/nanoarrow_device.h>
+
+#include <memory>
+#include <utility>
+
+namespace cudf {
+
+struct arrow_array_container {
+  arrow_array_container() = default;
+
+  template <typename T>
+  arrow_array_container(ArrowSchema* schema_,
+                        T input_,
+                        rmm::cuda_stream_view stream,
+                        rmm::device_async_resource_ref mr)
+  {
+    ArrowSchemaMove(schema_, &schema);
+    auto output = cudf::to_arrow_device(std::move(input_), stream, mr);
+    ArrowDeviceArrayMove(output.get(), &owner);
+  }
+
+  arrow_array_container(ArrowSchema const* schema_,
+                        ArrowDeviceArray* input_,
+                        rmm::cuda_stream_view stream,
+                        rmm::device_async_resource_ref mr)
+  {
+    switch (input_->device_type) {
+      case ARROW_DEVICE_CUDA:
+      case ARROW_DEVICE_CUDA_HOST:
+      case ARROW_DEVICE_CUDA_MANAGED: {
+        ArrowSchemaDeepCopy(schema_, &schema);
+        auto& device_arr = owner;
+        ArrowArrayMove(&input_->array, &device_arr.array);
+        device_arr.device_type = input_->device_type;
+        // Pointing to the existing sync event is safe because the underlying
+        // event must be managed by the private data and the release callback.
+        device_arr.sync_event = input_->sync_event;
+        device_arr.device_id  = input_->device_id;
+        break;
+      }
+      default: throw std::runtime_error("Unsupported ArrowDeviceArray type");
+    }
+  }
+  ~arrow_array_container()
+  {
+    if (owner.array.release != nullptr) { ArrowArrayRelease(&owner.array); }
+  }
+
+  ArrowDeviceArray owner{};  //< ArrowDeviceArray that owns the data
+  ArrowSchema schema{};      //< ArrowSchema that describes the data
```

Comment:
any reason to not just use `nanoarrow::UniqueSchema` and friends?

Reply:
I originally wasn't sure if I would wind up with such a clear separation between the user-facing types and the underlying container, so I was worried about exposing nanoarrow types in the public interface. Now that everything's using PImpl for the container that is moot, so I can probably switch over to nanoarrow types here.